include/net/cloud.h

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#ifndef ZEPHYR_INCLUDE_CLOUD_H_
+#define ZEPHYR_INCLUDE_CLOUD_H_
+
+/* Cloud API */
+
+#include <zephyr.h>
+
+/**@brief Cloud backend states. */
+enum cloud_state {
+	CLOUD_STATE_DISCONNECTED,
+	CLOUD_STATE_DISCONNECTING,
+	CLOUD_STATE_CONNECTED,
+	CLOUD_STATE_CONNECTING,
+	CLOUD_STATE_BUSY,
+	CLOUD_STATE_ERROR,
+	CLOUD_STATE_COUNT
+};
+
+/**@brief Cloud events that can be notified asynchronously by the backend. */
+enum cloud_event_type {
+	CLOUD_EVT_CONNECTED,
+	CLOUD_EVT_DISCONNECTED,
+	CLOUD_EVT_READY,
+	CLOUD_EVT_ERROR,
+	CLOUD_EVT_DATA_SENT,
+	CLOUD_EVT_DATA_RECEIVED,
+	CLOUD_EVT_COUNT
+};
+
+/**@brief Quality of Service for message sent by a cloud backend. */
+enum cloud_qos {
+	CLOUD_QOS_AT_MOST_ONCE,
+	CLOUD_QOS_AT_LEAST_ONCE,
+	CLOUD_QOS_EXACTLY_ONCE,
+	CLOUD_QOS_COUNT
+};
+
+/**@brief Quality of Service for message sent by a cloud backend. */
+enum cloud_resource {
+	CLOUD_EP_TOPIC,
+	CLOUD_EP_URI,
+	CLOUD_EP_COUNT
+};
+
+/* Forward declaration of cloud backend type. */
+struct cloud_backend;
+
+/**@brief Cloud message type. */
+struct cloud_msg {
+	char *payload;
+	size_t len;
+	enum cloud_qos qos;
+	struct {
+		enum cloud_resource type;
+		char *str;
+		size_t len;
+	} resource;
+};
+
+/**@brief Cloud event type. */
+struct cloud_event {
+	enum cloud_event_type type;
+	union {
+		struct cloud_msg msg;
+		int err;
+	} data;
+};
+
+/**
+ * @brief Cloud event handler function type.
+ *
+ * @param[in] backend Pointer to cloud backend.
+ * @param[in] evt Pointer to cloud event.
+ */
+typedef void (*cloud_evt_handler_t)(const struct cloud_backend *const backend,
+				    const struct cloud_event *const evt);
+
+/**
+ * @brief Cloud backend API.
+ *
+ * All cloud backends must expose this API.
```

Comment:
Which of those are mandatory? I see that ping is not.

Reply:
Initially I was thinking all except `ping`. Should be stated in the docs